src/components/UIInventory.cpp

```diff
@@ -0,0 +1,227 @@
+#include "UIInventory.hpp"
+#include <GUI/BsGUIButton.h>
+#include <GUI/BsGUILabel.h>
+#include <GUI/BsGUILayoutY.h>
+#include <GUI/BsGUIPanel.h>
+#include <GUI/BsGUIScrollArea.h>
+#include <RTTI/RTTI_UIInventory.hpp>
+#include <components/Inventory.hpp>
+#include <exception/Throw.hpp>
+#include <log/logging.hpp>
+
+namespace REGoth
+{
+  UIInventory::UIInventory(const bs::HSceneObject& parent, HUIElement parentUiElement)
+      : UIElement(parent, parentUiElement, new bs::GUIPanel())
+  {
+    setName("UIInventory");
+
+    mScrollArea = layout().addNewElement<bs::GUIScrollArea>(bs::ScrollBarType::ShowIfDoesntFit,
+                                                            bs::ScrollBarType::NeverShow);
+  }
+
+  UIInventory::~UIInventory()
+  {
+  }
+
+  void UIInventory::setViewedInventory(HInventory inventory)
+  {
+    if (mViewedInventory)
+    {
+      // FIXME: There might be other code registering a callback here, but for some reason
+      //        we can only clear the whole thing! This should be handled differently, but
+      //        I don't know how yet.
+      mViewedInventory->OnItemChanged.clear();
+
+      REGOTH_LOG(Warning, Uncategorized, "[UIInventory] Cleared Inventory OnItemChanged callback!");
+    }
+
+    mViewedInventory = inventory;
+
+    if (mRegisteredOnItemChangedEvent)
+    {
+      mRegisteredOnItemChangedEvent.disconnect();
+    }
+
+    if (mViewedInventory)
+    {
+      mRegisteredOnItemChangedEvent =
+          mViewedInventory->OnItemChanged.connect([this](const bs::String& instance) {
+            // Get notified when the viewed inventory changes
+            onInventoryItemUpdated(instance);
+          });
+
+      forceUpdateAll();
+    }
+  }
+
+  void UIInventory::onInventoryItemUpdated(const bs::String& instance)
+  {
+    bs::UINT32 countUI        = itemCount(instance);
+    bs::UINT32 countInventory = mViewedInventory->itemCount(instance);
+
+    if (countUI < countInventory)
+    {
+      addItem(instance, instance, countInventory - countUI);
+    }
+    else if (countUI > countInventory)
+    {
+      removeItem(instance, countUI - countInventory);
+    }
+    else
+    {
+      // Count is the same, do nothing
```

Comment:
Any specific cases were this can happen?
This would only happen when calling giveItem or removeItem with count of 0.

Reply:
That could also mean that the items name or value was changed as this is a general "something happened to that item" callback. Right now the only usecase is to update the count, but there might be more in the future.

Adding the else here is just defensive programming.

Reply:
Makes sense. I only thought about what happens right now.